js/bigone.js

```diff
@@ -200,6 +200,15 @@ module.exports = class bigone extends Exchange {
                 'quoteId': quoteId,
                 'type': 'spot',
                 'spot': true,
+                'margin': false,
+                'future': false,
+                'swap': false,
+                'linear': undefined,
+                'inverse': undefined,
+                'expiry': undefined,
+                'expiryDatetime': undefined,
+                'contract': false,
+                'contractSize': undefined,
```

Comment:
Could you add these fields also?

```
'settle': undefined,
'settleId': undefined,
'option': false,
'strike': undefined,
'optionType': undefined,
'limits': {
    'leverage': {
        'min': undefined,
        'max': undefined,
    },
```

Reply:
yep i'll add settle and settleId to these exchanges, however i'd wait on adding strike and optionType everywhere yet since we haven't even started to unify options yet

Reply:
so far we have support for them on one exchange (deribit) and even that support is shit

Reply:
instead of just guessing how to unify options we should take our time to make a few implementations and trade on a few options exchanges and see what we can do

Reply:
We have to have `option` and `optionType` for options, and we already have that in okex and in a few other exchanges, huobi is introducing usdc options, and we also have leverage limits as well. So, I think, it won't hurt to also have `strike` there.

Reply:
We include the strike in the symbol name for options because it's needed to differentiate between option markets, so I believe its needed them always

Reply:
> we already have that in okex and in a few other exchanges

We have it in quite a few exchanges actually, almost 30 I think

Reply:
i suggest to not bother at all with options until we finish with futures since we will probably need to re-edit everything, the options volume is tiny and probably neither of you have very much experience trading crypto options

Reply:
I think we should handle options if we encounter them along the way. They're not any better or worse than any other type of market. And since many exchanges have options, it is just a parallel vector to unification.

Reply:
At this point we know the minimal set of fields that will be needed for options, as @samgermain posted above.

Reply:
i don't disagree to add options, the point is to add it you cannot just guess, probably this is the revelant data, you need to actually trade options to find out what is important and how to format it

Reply:
I am pretty sure on that set of fields for options, it's not a guess, i've taken a look.

Reply:
right now only okex supports options in ccxt (more than just loading markets), so i think it would be cleaner if we finish adding futures, without options support everywhere, and then start to research how to add options every if necessary (which it may not be since the volumes are very small)

Reply:
@Dan-krm Knows options really well by the way

Reply:
crypto options aren't so popular tho

Reply:
the point isn't to support everything, but to support a subset that is used and make the code good, cause rn our futures support is still very ugly

Reply:
It's not a question of focusing on the unification of options. It's a simpler question of adding the above fields to the market structure, cause we already know we will need them.

Reply:
> @Dan-krm Knows options really well by the way

the question is whether to add it to the lib at this point

Reply:
> It's not a question of focusing on the unification of options. It's a simpler question of adding the above fields to the market structure, cause we already know we will need them.

I don't even think we will need them tbh, almost nobody is trading crypto options outside of deribit

Reply:
I say we should add those fields along with the unification of market type, which is what we are doing right now.

Reply:
and finally nobody will use ccxt to trade options and they will build there own code, i'd rather have 0% support for options than 10% support

Reply:
We already have a greater-than-0% support for options in the lib.

Reply:
there's not really too much point to have options inside of the markets when you cannot get the options orderbook using ccxt or options tickers or create options orders, i think it will be cleaner if we 

1. remove the options support from the markets
2. stop loading options markets

once we are ready and have done more research and we have more experience with options we should try to add them in their entirely not this half hearted bullshit support that we have for them rn

Reply:
CCXT has been evolving in this manner for a long while, so I don't see why we should necessarily dump options rn. To me adding the fields in question is much easier than removing the support for options.

> there's not really too much point to have options inside of the markets when you cannot get the options orderbook using ccxt or options tickers or create options orders

First, have you tried? Some functionality is working ok. Some bits are missing, which is also ok – we will add them. Also, we still list out all their implicit endpoints, including the options endpoints, anyway. So, I don't see why we should make any exception for options at all.

Reply:
<img width="486" alt="Screenshot 2022-01-15 at 17 14 30" src="https://user-images.githubusercontent.com/14319357/149637937-6c12d2a5-c053-4480-995b-0f2b6d9fabfd.png">

moreover with these volumes there is nothing to unify

Reply:
With those volumes there's at least market types and symbols to unify.

Reply:
> First, have you tried? Some functionality is working ok. Some bits are missing, which is also ok – we will add them. Also, we still list out all their implicit endpoints, including the options endpoints, anyway. So, I don't see why we should make any exception for options at all.

by some functionality is working you mean there is extremely limited functionality on just okex and deribit

Reply:
the point is that time is valuable, so by not supporting options we save time and simplify our codebase

Reply:
for the time being the options volumes are too centralised on deribit and there is no other exchange with over 10m volume

Reply:
It is not a question of unifying options everywhere in the lib, it's not even a question about options themselves, it's more about properly categorizing the markets. To me what you're suggesting is to leave this aspect unfinished.

Reply:
i'm not suggesting to leave it unfinished, rather to have the understanding that we're better off leaving it until a later date when the crypto options markets are more matured

Reply:
imo right now the options code is unfinished

Reply:
We will not delist Deribit at this point, so we will have to have those fields.

Reply:
The question to you is: since you're editing exactly that, could you also add those fields along the way or should we necessarily run multiple rounds of edits on the same lines done by different devs.

Reply:
> right now only okex supports options in ccxt (more than just loading markets), so i think it would be cleaner if we finish adding futures, without options support everywhere, and then start to research how to add options everywhere if necessary (which it may not be since the volumes are very small)

I feel like it would be more worthwhile to add options support on the big exchanges, rather adding futures support across every small exchange on CCXT

`equonex` has only 2 swap markets for example. To me that doesn't seem very worthwhile to complete before binance options for example, other than for the sake of claiming complete unification, which seems like it might be a task that would require more effort than it's worth. Right now the exchange capabilities say `3075 implemented, 281 emulated, 305 inexistentApi, 6860 notImplemented`, so there's more than twice as many not implemented than implemented. Lately I've been trying to unify just the fetchMarkets across every exchange, and that alone is a really long task.

I think that rather than aiming for complete unification across all 117 exchanges, it might be better to prioritize a list that includes the top exchanges and certified exchanges, and unifying those, but exchanges which don't fall on that list wouldn't get priority over adding options on the big exchanges.

I just think there's a lot more value in adding features that people want and that they're going to use over trying to unify every method across every exchange.

Please voice your opinion if you disagree though, I'm fine with just focusing on unification.

Reply:
we don't need to delist deribit, but i wouldn't add it to all the other exchanges

Reply:
> other than for the sake of claiming complete unification, which seems like it might be a task that would require more effort than it's worth

Complete unification does not exist. It's a myth.

Reply:
@samgermain I disagree because we're chasing volumes at ccxt and there is significantly more futures volumes on small exchanges than there is options volumes on big exchanges

Reply:
>  Lately I've been trying to unify just the fetchMarkets across every exchange, and that alone is a really long task.

It's not that long if you do it quickly

Reply:
> I agree, complete unification does not exist. It's a myth. CCXT is always in an incomplete state, cause there's always more and more things to unify.

yep, i don't disagree with this, which is why i'm arguing to not add those fields everywhere just yet

Reply:
@frosty00 

Currently there's about 30 exchanges that include the fields, so to unify the fetchMarkets method, we either need to remove it from those 30, or add it to the rest of the exchanges. It doesn't make a lot of sense to me to remove it because we know that we'll need to fields for options, which will get more popular over time

```
'option': false,
'strike': undefined,
'optionType': undefined,
```

Reply:
> we don't need to delist deribit, but i wouldn't add it to all the other exchanges

We also have options on okex, bybit, gateio, delta exchange.

Reply:
crypto options are also in the infant stage right now. There's not a lot of volume now, but it just started, just give it a year or two and it will be bigger, and it's not like it's a lot of work to add the fields `optionType` and `strike` to `fetchMarkets` or that it would cause any downsides, but it would make CCXT more prepared for the future

Reply:
> @frosty00
> 
> Currently there's about 30 exchanges that include the fields
> 
> ```
> 'option': false,
> 'strike': undefined,
> 'optionType': undefined,
> ```
> 
> so to unify the fetchMarkets method, we either need to remove it from those 30, or add it to the rest of the exchanges. It doesn't make a lot of sense to me to remove it because we know that we'll need to fields for options, which will get more popular over time

firstly it is only in 15 exchanges.

Secondly, this is what I'm saying, we should remove these fields on all those exchanges except the ones where we have support for options for now until the options mature further, since the vast majority of exchanges have 0 support for options. So in a way those fields will be optional (pardon the pun)

Reply:
What would be the benefit in removing it? This would make the `fetchMarkets` response less unified

Reply:
@frosty00 we need to make the markets from all exchanges have the same predictable structure, though, otherwise it spaghettifies the code in the userland with unnecessary checks.

Reply:
> @frosty00 we need to make the markets from all exchanges have the same predictable structure, though. Otherwise it spaghettifies the code in the userland with unnecessary checks.

[For example](https://github.com/ccxt/ccxt/pull/11302#discussion_r784446357)

Reply:
the reason for removing it is that less than 10% of exchanges support options so we don't really need to have unnecessary fields inside of `.markets`, if options become more popular then i would add to all the exchanges. with regard to the unnecessary checks it won't be a problem since nobody will attempt to trade options on an exchange that doesn't have an option market

Reply:
> if options become more popular then i would add to all the exchanges

They are already becoming more popular, more and more exchanges are introducing them, and now after binance the others will follow even faster.

I cannot really imagine CCXT in the longer run with everything-but-options, so to me removing this bit now just to add it later still makes less sense than keeping it and adding it everywhere along the way. We don't even need those fields for the purpose of unifying options-trading yet, we only need them to unify markets.

Reply:
we can add these three fields then, the cost is a only couple of unused fields

Reply:
https://www.coindesk.com/markets/2021/12/02/goldman-sachs-sees-crypto-options-markets-as-next-big-step-for-institutional-adoption/

Reply:
> we can add these three fields then, the cost is a only couple of unused fields

lmk if you're going to amend these PRs, or if i should do it upon merging

Reply:
i'll amend these two

Reply:
> > other than for the sake of claiming complete unification, which seems like it might be a task that would require more effort than it's worth
> 
> I agree, complete unification does not exist. It's a myth. CCXT is always in an incomplete state, cause there's always more and more things to unify.

I think that we should try to completely unify all the certified exchanges though, with all the unified methods being `true` or `false` within has, rather than undefined. That way the certified exchanges are "complete"

Reply:
> I think that we should try to completely unify all the certified exchanges though, with all the unified methods being `true` or `false` within has, rather than undefined. That way the certified exchanges are "complete"

I agree with that.